incubator/hnc/pkg/stats/stats.go

```diff
@@ -0,0 +1,90 @@
+package stats
+
+import (
+	"strconv"
+	"sync/atomic"
+	"time"
+
+	"github.com/go-logr/logr"
+	ctrl "sigs.k8s.io/controller-runtime"
+)
+
+type counter int32
+
+func (c *counter) Incr() {
+	i := int32(*c)
+	atomic.AddInt32(&i, 1)
+	*c = counter(i)
+}
+
+func (c *counter) Decr() {
+	i := int32(*c)
+	atomic.AddInt32(&i, -1)
+	*c = counter(i)
+}
+
+type stats struct {
+	// HcTot is the total number of HierarchyConfig reconciliations.
+	HcTot counter
+
+	// HcCur is the currently undergoing number of HierarchyConfig reconciliations.
+	HcCur counter
+
+	// ObTot is the total number of object reconciliations.
+	ObTot counter
+
+	// ObCur is the currently undergoing number of object reconciliations.
+	ObCur counter
+
+	// HcRead is the total number of HierarchyConfig reads.
+	HcRead counter
+
+	// HcWrite is the total number of HierarchyConfig writes.
+	HcWrite counter
+
+	// ObRead is the total number of object reads.
+	ObRead counter
+
+	// ObWrite is the total number of object writes.
+	ObWrite counter
+}
+
+var Stats stats
+
+// LogActivity generates logs for performance testing.
+func LogActivity() {
+	log := ctrl.Log.WithName("reconcileCounter")
+	var total, cur counter = 0, 0
+	working := false
+	actN := 1
+	go func() {
+		// run forever
+		for {
+			// Log activity only when the controllers were still working in the last 0.5s.
+			time.Sleep(500 * time.Millisecond)
+			if Stats.HcTot+Stats.ObTot != total || cur != 0 {
+				// If the controller was previously idle, change its status and log it's started.
+				if working == false {
+					working = true
+					msg := "Activity-" + strconv.Itoa(actN) + "-started"
+					logActivity(log, msg, Stats)
+				}
+				logActivity(log, "Activity", Stats)
```

Comment:
I'm wondering if "activity" should be one of the structured fields in the log, and then also have a "status" field that can be either "start," "continue" or "finish". That'll make the logs more uniform and easier to grep, I think.

Reply:
In the performance testing scripts, since the logs are grep by intervals (e.g. 5s), the activity start and finish have to be numbered to be able to map to the testing actions. Currently it's grep by e.g. "Activity-$i-finished".

Do you mean generating "activity" number the same way as generating "rid"? It does look more uniform. I will make the changes in this first and figure out how to grep the log in the next PR?

Reply:
Yeah, pretty much. You could probably do a grep like:

```
cat log | grep "\"activity\": \"5\"" | grep "\"status\": \"finished\"
```

which isn't really as pretty but is definitely more in keeping with how structured logs usually work. wdyt?